Support for custom purgers

lstrojny opened the pull request:
This PR allows users to specify a custom purger (as previously requested in #116).

As a user, I can register my custom purger factory, that implements `createForEntityManager(?string $emName, EntityManagerInterface $em, array $excluded = [], bool $purgeWithTruncate = false) : PurgerInterface` with the tag `doctrine.fixtures.purger_factory` and some `alias` and then run the command with `--purger=$alias`. If the `ORMPurger` is used `--purge-exclusions`, a list of tables that ought to not be purged, are supported.

lstrojny committed:
Fix expected error message in integration test

lstrojny committed:
Allow specifying custom purgers

Custom purgers can now specified by implementing a PurgerFactory and
registering it with a tag "doctrine.fixtures.purger_factory" and
specifying an alias.

Co-Authored-By: Claudio Zizza <859964+SenseException@users.noreply.github.com>

greg0ire commented on Command/LoadDataFixturesDoctrineCommand.php:
```diff
@@ -28,7 +29,10 @@ class LoadDataFixturesDoctrineCommand extends DoctrineCommand
+    /** @var ContainerInterface */
+    private $purgerFactories;
+
+    public function __construct(SymfonyFixturesLoader $fixturesLoader, ContainerInterface $purgerFactories, ?ManagerRegistry $doctrine = null)
```
Please avoid the BC break

lstrojny replied:
Default factories to null, current behavior if null?

greg0ire commented:
Please add some docs

lstrojny commented:
> Please add some docs

Done

greg0ire commented on Command/LoadDataFixturesDoctrineCommand.php:
```diff
@@ -28,7 +29,10 @@ class LoadDataFixturesDoctrineCommand extends DoctrineCommand
     private $fixturesLoader;
 
-    public function __construct(SymfonyFixturesLoader $fixturesLoader, ?ManagerRegistry $doctrine = null)
+    /** @var ContainerInterface */
```
```suggestion
    /** @var ContainerInterface|null */
```

greg0ire commented on Command/LoadDataFixturesDoctrineCommand.php:
```diff
@@ -118,8 +125,12 @@ protected function execute(InputInterface $input, OutputInterface $output)
-        $purger = new ORMPurger($em);
-        $purger->setPurgeMode($input->getOption('purge-with-truncate') ? ORMPurger::PURGE_MODE_TRUNCATE : ORMPurger::PURGE_MODE_DELETE);
+        $purger = $this->purgerFactories
+            ? $this->purgerFactories->get($input->getOption('purger'))->createForEntityManager($input->getOption('em'), $em, $input->getOption('purge-exclusions'))
```
```suggestion
            ? $this->purgerFactories->get($input->getOption('purger'))->createForEntityManager(
                $input->getOption('em'),
                $em,
                $input->getOption('purge-exclusions')
            )
```

greg0ire commented on Command/LoadDataFixturesDoctrineCommand.php:
```diff
@@ -118,8 +125,12 @@ protected function execute(InputInterface $input, OutputInterface $output)
+            ? $this->purgerFactories->get($input->getOption('purger'))->createForEntityManager($input->getOption('em'), $em, $input->getOption('purge-exclusions'))
+            : new ORMPurger($em, $input->getOption('purge-exclusions'));
+        if ($purger instanceof ORMPurger) {
+            $purger->setPurgeMode($input->getOption('purge-with-truncate') ? ORMPurger::PURGE_MODE_TRUNCATE : ORMPurger::PURGE_MODE_DELETE);
```
```suggestion
            $purger->setPurgeMode(
                $input->getOption('purge-with-truncate') ?
                ORMPurger::PURGE_MODE_TRUNCATE :
                ORMPurger::PURGE_MODE_DELETE
            );
```

greg0ire commented on Resources/doc/index.rst:
```diff
@@ -293,3 +293,93 @@ fixture using the ``UserFixtures`` group:
+Specifying purging behavior
+---------------------------
+
+By default alls previously existing data is purged using ``DELETE FROM table`` statements. If you prefer to use
```
```suggestion
By default all previously existing data is purged using ``DELETE FROM table`` statements. If you prefer to use
```

greg0ire commented on Resources/doc/index.rst:
```diff
@@ -293,3 +293,93 @@ fixture using the ``UserFixtures`` group:
+
+    class CustomPurgerFactory implements PurgerFactory
+    {
+        public function createForEntityManager(?string $emName, EntityManagerInterface $em, array $excluded = []) : PurgerInterface;
```
```suggestion
        public function createForEntityManager(?string $emName, EntityManagerInterface $em, array $excluded = []): PurgerInterface;
```

1. to be consistent with the style you used above
2. we are going to remove that space in the next version of our coding standard.

greg0ire commented on Resources/doc/index.rst:
```diff
@@ -293,3 +293,93 @@ fixture using the ``UserFixtures`` group:
+        }
+    }
+
+The next step is to register our custom purger factory and specify it’s alias.
```
```suggestion
The next step is to register our custom purger factory and specify its alias.
```

greg0ire commented on Resources/doc/index.rst:
```diff
@@ -293,3 +293,93 @@ fixture using the ``UserFixtures`` group:
+
+        use App\Purger\CustomerPurgerFactory;
+
+        return function(ContainerConfigurator $configurator) {
```
```suggestion
        return function(ContainerConfigurator $configurator): void {
```

greg0ire commented on Tests/IntegrationTest.php:
```diff
@@ -313,7 +478,7 @@ class IntegrationTestKernel extends Kernel
     public function __construct(string $environment, bool $debug)
     {
-        $this->randomKey = rand(100, 999);
+        $this->randomKey = hash('sha3-384', spl_object_hash($this) . self::$invocations++);
```
What is this about?

lstrojny replied:
When running the suite before, 1 out of 1000 times I got weird errors, because it reused a previous container. This makes sure that this never happens

greg0ire replied:
Please move this to a separate commit that will contain this explanation, and take it as an occasion to squash everything else into a single commit.

greg0ire commented:
Please have a look at the test suite, it fails.

lstrojny committed:
Make integration tests more robust

When running the suite in 1 out of 1000 times I a previous container was incorrectly reused

greg0ire reviewed (approved):
Great job with the commits!

lstrojny commented:
Still working on solving the deprecation warnings

greg0ire commented on Command/LoadDataFixturesDoctrineCommand.php:
```diff
@@ -28,19 +31,27 @@ class LoadDataFixturesDoctrineCommand extends DoctrineCommand
+                DeprecatedManagerRegistry::class
             ), E_USER_DEPRECATED);
+        } elseif (!$doctrine instanceof ManagerRegistry && !$doctrine instanceof DeprecatedManagerRegistry) {
+            throw new TypeError(sprintf('Argument 2 passed to %s() must implement interface %s or %s', __METHOD__, ManagerRegistry::class, DeprecatedManagerRegistry::class));
```
```suggestion
            throw new TypeError(sprintf(
                'Argument 2 passed to %s() must implement interface %s or preferably %s',
                __METHOD__,
                DeprecatedManagerRegistry::class,
                ManagerRegistry::class
            ));
```

lstrojny replied:
Done

lstrojny commented:
Alright, here we are:
 - Fixing the manager deprecation properly required a bit more work
 - Commits are clean and atomic now
 - Build passes (https://travis-ci.org/github/doctrine/DoctrineFixturesBundle/builds/668042364)
 - Docs exists

I guess we are ready to go

Thanks for the collaboration, @greg0ire

lstrojny commented:
@greg0ire meditating over it a bit, I've moved more responsibilities into the factory and added a warning of a user passes an unknown purger so that we are not falling back to the default factory silently.

lstrojny commented:
Tested this in a project and it works. Found and fixed another bug (truncate mode was not correctly passed).

SenseException commented on Resources/doc/index.rst:
```diff
@@ -293,3 +293,93 @@ fixture using the ``UserFixtures`` group:
+    // ...
+    class CustomPurger implements PurgerInterface
+    {
+        public function purge(): void
```
```suggestion
        public function purge() : void
```

SenseException commented on Resources/doc/index.rst:
```diff
@@ -293,3 +293,93 @@ fixture using the ``UserFixtures`` group:
+
+    class CustomPurgerFactory implements PurgerFactory
+    {
+        public function createForEntityManager(?string $emName, EntityManagerInterface $em, array $excluded = [], bool $purgeWithTruncate = false): PurgerInterface;
```
```suggestion
        public function createForEntityManager(?string $emName, EntityManagerInterface $em, array $excluded = [], bool $purgeWithTruncate = false) : PurgerInterface;
```

SenseException commented on Resources/doc/index.rst:
```diff
@@ -293,3 +293,93 @@ fixture using the ``UserFixtures`` group:
+
+        use App\Purger\CustomerPurgerFactory;
+
+        return function(ContainerConfigurator $configurator): void {
```
```suggestion
        return function(ContainerConfigurator $configurator) : void {
```

SenseException reviewed (commented):
Just a few small return type format changes in the docs.

lstrojny commented:
@SenseException done

greg0ire commented:
Thanks @lstrojny !
